### PR TITLE
ENSURE loss

### PR DESCRIPTION
**Pull request by Andrewwango**
ENSURE loss from https://arxiv.org/abs/2010.10631

### Checks to be done before submitting your PR
- [ ] `python3 -m pytest tests/` runs successfully.
- [ ] `black .` runs successfully.
- [ ] `make html` runs successfully (in the `docs/` directory).
- [ ] Updated docstrings related to the changes (as applicable).
- [ ] Added an entry to the [CHANGELOG.rst](../CHANGELOG.rst).

### CODE REVIEW

**Review by tachella — COMMENTED** [review]
I feel that this loss is a too specific version of the SURE loss, where the metric B is set using an expectation over the physics generator.

In my opinion this would clutter a bit the library and confuse practitioners, so I'm not sure about adding it.

why not include it in an example?

**Comment by tachella** [comment]
The loss is too close to SURE to motivate a new class in my opinion.


I would nonetheless maybe keep a function for computing the metric based on a physics generator, e.g. as done in

```
d = physics_generator.step(batch_size=2000)["mask"].mean(0, keepdim=True)
self.dsqrti = 1 / d.sqrt()
```

but otherwise I would suggest to close the PR

File: deepinv/loss/mri/sure.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+    .. math::
+
+        y \sim\mathcal{N}(u,\sigma^2 I) \quad \text{with}\quad u= A(x).
+
+    The loss is computed as
+
+    .. math::
+
+        \frac{1}{m}\|W^{-1}P(A^{\dagger}y - \inverse{y})\|_2^2 +\frac{2\sigma^2}{m\tau}b^{\top} \left(\inverse{A^{\top}y+\tau b_i} -
```
Comment (tachella) [inline deepinv/loss/mri/sure.py:31]: why \inverse{A^{\top}y} ? can't this loss work for a general network R(y,A)?

If that's the case, we should specify that it only works for Artifact Removal networks?

File: deepinv/loss/mri/sure.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        r"""
+        Monte-Carlo estimation for the divergence of f(x).
+
+        :param torch.Tensor x_net: Reconstructions.
+        :param torch.Tensor y: Measurements.
+        :param deepinv.physics.Physics physics: Forward operator associated with the measurements.
+        :param torch.nn.Module f: Reconstruction network.
+        """
+        b = torch.empty_like(y).normal_(generator=self.rng)
+        x2 = f(physics.A(physics.A_adjoint(y) + b * self.tau), physics)
```
Comment (tachella) [inline deepinv/loss/mri/sure.py:75]: this doesn't exactly match the math definition above

**Review by tachella — CHANGES_REQUESTED** [review]

File: deepinv/loss/mri/sure.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+    def __init__(
+        self,
+        sigma: float,
+        physics_generator: PhysicsGenerator,
+        tau: float = 0.01,
+        rng: torch.Generator = None,
+    ):
+        super().__init__(sigma=sigma, tau=tau, rng=rng)
+        d = physics_generator.step(batch_size=2000)["mask"].mean(0, keepdim=True)
```
[the remark this answers is not in the document]
Comment (Andrewwango) [reply deepinv/loss/mri/sure.py]: Yes potentially, and I raise this issue here: https://github.com/deepinv/deepinv/issues/486

[the remark this answers is not in the document]
Comment (Andrewwango) [reply deepinv/loss/mri/sure.py]: But for now it should be ok, and I propose a `physics_generator.average`

File: deepinv/loss/mri/sure.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+    The loss is computed as
+
+    .. math::
+
+        \frac{1}{m}\|W^{-1}P(A^{\dagger}y - \inverse{y})\|_2^2 +\frac{2\sigma^2}{m\tau}b^{\top} \left(\inverse{A^{\top}y+\tau b_i} -
+        \inverse{A^{\top}y}\right)
+
+    where :math:`R` is the trainable network, :math:`A` is the forward operator,
+    :math:`y` is the noisy measurement vector of size :math:`m`, :math:`W^{-1}P` is a weighted projection operator determined by the set of measurement operators,
```
[the remark this answers is not in the document]
Comment (Andrewwango) [reply deepinv/loss/mri/sure.py]: done

File: deepinv/loss/mri/sure.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    from deepinv.physics.generator.base import PhysicsGenerator
+    from deepinv.models.base import Reconstructor
+    from deepinv.physics.forward import Physics
+
+
+class ENSURELoss(SureGaussianLoss):
+    r"""
+    ENSURE loss for image reconstruction in Gaussian noise.
+
+    The loss function extends :class:`deepinv.loss.SureGaussianLoss` and is designed for the following noise model:
```
[the remark this answers is not in the document]
Comment (Andrewwango) [reply deepinv/loss/mri/sure.py]: done

File: deepinv/loss/mri/sure.py [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+    .. math::
```
[the remark this answers is not in the document]
Comment (Andrewwango) [reply deepinv/loss/mri/sure.py:23]: done

File: deepinv/loss/mri/sure.py (hunk 1 above)
Comment (Andrewwango) [reply deepinv/loss/mri/sure.py]: They state this is the case in their eq. 15
image
This is valid for ArtifactRemoval, unrolled etc.

Comment (Andrewwango) [reply deepinv/loss/mri/sure.py]: See below comment on how we generalise this

File: deepinv/loss/mri/sure.py (hunk 2 above)
Comment (Andrewwango) [reply deepinv/loss/mri/sure.py:78]: Firstly, we note that ENSURE only applies for networks $R(\cdot)=r(A^\top\cdot)$ such that $R: u\rightarrow x,u=A^\top y$ as noted in the above comment. Assume that the user is using a valid model.

To be compatible with our model framework $model: y\rightarrow x$, we use the fact that $A^\top A A^\top = A^\top$ and add the `physics.A`.

Comment (Andrewwango) [reply deepinv/loss/mri/sure.py:80]: Would you agree that this is a sufficiently general way of doing this, as long as we assume that the model that the user passes in is indeed in the form $r(A^\top\cdot)$?

Comment (Andrewwango) [reply deepinv/loss/mri/sure.py:80]: Would you agree that this is a sufficiently general way of doing this, as long as we assume that the model that the user passes in is indeed in the form $r(A^\top\cdot)$?

Comment (tachella) [reply deepinv/loss/mri/sure.py:80]: this also assumes that A^{t}AA^{t} = A^t which is only true for operators with binary eigenvalues

File: deepinv/loss/mri/sure.py (hunk 1 above)
Comment (tachella) [reply deepinv/loss/mri/sure.py]: they require in the proof that $\hat{\rho}$ (the reconstruction net) is independent of $s$ (aka $A_s$)
image

Comment (Andrewwango) [reply deepinv/loss/mri/sure.py]: I've addressed this in the docs and by adding an artifactremoval switch

File: deepinv/loss/mri/sure.py (hunk 2 above)
Comment (Andrewwango) [reply deepinv/loss/mri/sure.py]: I've addressed this in the docs and by adding an artifactremoval switch

File: deepinv/loss/mri/sure.py [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+    :math:`b\sim\mathcal{N}(0,I)`, :math:`\tau\geq 0` is a hyperparameter controlling the
+    Monte Carlo approximation of the divergence, and :math:`\Beta=W^{-1}P`
+    where :math:`W` is a weighting determined by the set of measurement operators and :math:`P` is the projection operator onto the range space of :math:`\A^\top`.
+
+    The ENSURE loss was proposed in `Aggarwal et al <https://arxiv.org/abs/2010.10631>`_ for MRI.
+
+    .. warning::
+
+        This loss was originally proposed only to be used with :class:`artifact removal models <deepinv.models.ArtifactRemoval>` which can be written in the form :math:`\inverse{\cdot}=r(A^\top\cdot)`.
+        If an artifact removal model is not used, then we input to network instead :math:`A A^\top\cdot` (assuming that :math:`A A^\top=I`).
```
Comment (tachella) [inline deepinv/loss/mri/sure.py:48]: sorry I'm still bugging on this, but I would evaluate a general model as `model(y,physics)` instead of adding :math:`A A^\top\cdot`

File: deepinv/loss/mri/sure.py [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+
+    .. math::
+
+        \frac{1}{m}\|\Beta(A^{\dagger}y - \inverse{y})\|_2^2 +\frac{2\sigma^2}{m\tau}b^{\top} \left(\inverse{A^{\top}y+\tau b_i} -
+        \inverse{A^{\top}y}\right)
+
+    where :math:`R` is the trainable network (which takes :math:`A^\top y` as input),
+    :math:`A` is the forward operator,
+    :math:`y` is the noisy measurement vector of size :math:`m`,
+    :math:`b\sim\mathcal{N}(0,I)`, :math:`\tau\geq 0` is a hyperparameter controlling the
```
Comment (tachella) [inline deepinv/loss/mri/sure.py:39]: I would be more specific on how W is computed.

I think this only works for forward operators with an SVD decomposition, ie LinearPhysics, otherwise it is not feasible to compute W. We could clarify this too.

**Review by tachella — CHANGES_REQUESTED** [review]

File: deepinv/loss/mri/sure.py [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+        :param torch.Tensor x_net: Reconstructions.
+        :param torch.Tensor y: Measurements.
+        :param deepinv.physics.Physics physics: Forward operator associated with the measurements.
+        :param torch.nn.Module f: Reconstruction network.
+        """
+        b = torch.empty_like(y).normal_(generator=self.rng)
+        u = physics.A_adjoint(y) + b * self.tau
+
+        if isinstance(f, ArtifactRemoval):
+            x2 = f.backbone_inference(u, physics, None)
```
Comment (tachella) [inline deepinv/loss/mri/sure.py]: Sorry for going in circles, but I don't get why we need to do this special treatment here. Why not defining u=y + tau * b and then applying model(y, physics)?

Comment (Andrewwango) [reply deepinv/loss/mri/sure.py]: Because according to the paper, the model input is $A^\top y + b\tau $, not $A^\top(y+b\tau)$?

Comment (Andrewwango) [reply deepinv/loss/mri/sure.py]: I wanted to replicate the formulae from the paper rather than trying generalise

Comment (Andrewwango) [reply deepinv/loss/mri/sure.py]: i.e. figure 1b) of https://arxiv.org/abs/2010.10631

Comment (tachella) [reply deepinv/loss/mri/sure.py]: this is too hacky for me, I think it defeats the point of the library to have very hand-crafted solutions.

It will only work when `y` has the same size as `A_adjoint(y)`, and it is very weird to add `b` noise in the nullspace of `A`

I would go for the same divergence computation as in the SureGaussianLoss

Comment (Andrewwango) [reply deepinv/loss/mri/sure.py]: done

File: docs/source/api/deepinv.loss.rst [hunk 10]
```diff
@@ -67,6 +67,17 @@ Specialized self-supervised losses for MRI
...
+--------------------------
+.. userguide:: mri-losses
+
+.. autosummary::
+   :toctree: stubs
+   :template: myclass_template.rst
+   :nosignatures:
+
+    deepinv.loss.mri.ENSURELoss
+
```
Comment (tachella) [inline docs/source/api/deepinv.loss.rst:70]: Phase2Phase and Artifact2Artifact WeightedSplittingLoss RobustSplittingLoss shouldn't go here too?

File: deepinv/loss/mri/sure.py [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+    ) -> Tensor:
+        r"""
+        Monte-Carlo estimation for the divergence of f(x).
+
+        :param torch.Tensor x_net: Reconstructions.
+        :param torch.Tensor y: Measurements.
+        :param deepinv.physics.Physics physics: Forward operator associated with the measurements.
+        :param torch.nn.Module f: Reconstruction network.
+        """
+        return mc_div(physics(x_net), y, f, physics, tau=self.tau, rng=self.rng)
```
Comment (tachella) [inline deepinv/loss/mri/sure.py]:
```suggestion
        return mc_div(physics.A(x_net), y, f, physics, tau=self.tau, rng=self.rng)
```

File: deepinv/loss/mri/sure.py [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+
+    where :math:`R` is the trainable network (which takes :math:`A^\top y` as input),
+    :math:`A` is the forward operator,
+    :math:`y` is the noisy measurement vector of size :math:`m`,
+    :math:`b\sim\mathcal{N}(0,I)`, :math:`\tau\geq 0` is a hyperparameter controlling the
+    Monte Carlo approximation of the divergence, and :math:`\Beta=W^{-1}P`
+    where :math:`P` is the projection operator onto the range space of :math:`\A^\top`
+    and :math:`W` is a weighting determined by the set of measurement operators where :math:`W^2=\mathbb{E}\left[P\right]`.
+
+    The ENSURE loss was proposed in `Aggarwal et al <https://arxiv.org/abs/2010.10631>`_ for MRI.
```
Comment (tachella) [inline deepinv/loss/mri/sure.py]: use footcite for the citation

**Review by tachella — CHANGES_REQUESTED** [review]
final thing to fix

File: deepinv/loss/mri/sure.py (hunk 12 above)
Comment (Andrewwango) [reply deepinv/loss/mri/sure.py]: done

**Comment by tachella** [comment]
There's some black issues
